pkg/comp-functions/functions/common/resources.go

```diff
@@ -3,43 +3,111 @@ package common
 import (
 	vshnv1 "github.com/vshn/appcat/v4/apis/vshn/v1"
 	"github.com/vshn/appcat/v4/pkg/common/utils"
+	"k8s.io/apimachinery/pkg/api/resource"
 )
 
 type Resources struct {
-	ReqMem string
-	ReqCPU string
-	Mem    string
-	CPU    string
-	Disk   string
+	ReqMem resource.Quantity
+	ReqCPU resource.Quantity
+	Mem    resource.Quantity
+	CPU    resource.Quantity
+	Disk   resource.Quantity
 }
 
-func GetResources(size *vshnv1.VSHNSizeSpec, r utils.Resources) Resources {
-	reqMem := size.Requests.Memory
-	reqCPU := size.Requests.CPU
-	mem := size.Memory
-	cpu := size.CPU
-	disk := size.Disk
+func GetResources(size *vshnv1.VSHNSizeSpec, plan utils.Resources) (Resources, error) {
+	reqMem := resource.Quantity{}
+	reqCPU := resource.Quantity{}
+	mem := resource.Quantity{}
+	cpu := resource.Quantity{}
+	disk := plan.Disk
 
-	if reqMem == "" {
-		reqMem = r.MemoryRequests.String()
+	var err error
+
+	if size.Requests.Memory != "" {
+		reqMem, err = resource.ParseQuantity(size.Requests.Memory)
+		if err != nil {
+			return Resources{}, err
+		}
 	}
-	if reqCPU == "" {
-		reqCPU = r.CPURequests.String()
+
+	if size.Requests.CPU != "" {
+		reqCPU, err = resource.ParseQuantity(size.Requests.CPU)
+		if err != nil {
+			return Resources{}, err
+		}
 	}
-	if mem == "" {
-		mem = r.MemoryLimits.String()
+
+	if size.Memory != "" {
+		mem, err = resource.ParseQuantity(size.Memory)
+		if err != nil {
+			return Resources{}, err
+		}
 	}
-	if cpu == "" {
-		cpu = r.CPULimits.String()
+
+	if size.CPU != "" {
+		cpu, err = resource.ParseQuantity(size.CPU)
+		if err != nil {
+			return Resources{}, err
+		}
 	}
-	if disk == "" {
-		disk = r.Disk.String()
+
+	if size.Disk != "" {
+		disk, err = resource.ParseQuantity(size.Disk)
+		if err != nil {
+			return Resources{}, err
+		}
 	}
+
+	memLimit := getLimit(reqMem, mem, plan.MemoryLimits)
+	memReq := getRequest(reqMem, memLimit, plan.MemoryRequests)
+
+	cpuLimit := getLimit(reqCPU, cpu, plan.CPULimits)
+	cpuReq := getRequest(reqCPU, cpuLimit, plan.CPURequests)
+
 	return Resources{
-		ReqMem: reqMem,
-		ReqCPU: reqCPU,
-		Mem:    mem,
-		CPU:    cpu,
+		ReqMem: memReq,
+		ReqCPU: cpuReq,
+		Mem:    memLimit,
+		CPU:    cpuLimit,
 		Disk:   disk,
+	}, nil
+}
+
```

Comment:
please add a comment for the logic of this function, you can add it here or you could add in each `if else` clause.

Reply:
Done